docs/proposals/rebalancing-clusters-across-shards-dynamically.md

```diff
@@ -0,0 +1,144 @@
+
+The current sharding mechanism relies on predictable pod names for the application controller to determine which shard a given replica should impersonate, e.g. the first replica of the StatefulSet (argocd-application-controller-0) will be the first shard, the second replica (argocd-application-controller-1) will be the second and so forth. 
+
+## Motivation
+
+If the number of available shards is changed (i.e. one or more application controller replicas are added or removed), all pods in the statefulset have to be restarted so that the managed clusters are redistributed over the available shards. Additionally, the application controller workload is deployed as a StatefulSet, which is not intended for dynamic horizontal scaling.
+
+### Goals
+
+- Improve the application controller’s ability to scale horizontally with a growing number of clusters
+- Remove the need to run application controller as a StatefulSet workload
+
+### Non-Goals
+
+- Expand the scope of sharding to other assets than clusters (e.g. applications)
+- Make a single shard highly available (e.g. by having 2 or more replicas by shard)
+
+## Proposal
+
+### Why use Deployments instead of StatefulSet:
+StatefulSet is a Kubernetes resource that manages multiple pods that have unique identities, and are not interchangeable (unlike a regular Kubernetes Deployment, in which pods are stateless and can be destroyed and recreated as often as needed). 
+
+Stateless applications scale horizontally very easily as compared to stateful applications due to the fact that infrastructure allows adding as many computing resources as needed. Changing the StatefulSet to Deployments for Application Controller will allow us to dynamically scale the replicas without restarting existing application controller pods. Also, the shard to application controller assignment would help in making sure the shards are scaled and distributed across the available healhty replicas of application controllers.
+
+### Distributing shards among Application Controllers:
+
+Inorder to be able to accurately know which shards are being managed by which application-controller, especially in scenarios of redistribution of load, addition/removal of `application controller`, etc., we would need to have a mechanism to assign clusters to the shards. 
+
+The service account used by the application controller has read access to all the resources in the cluster. Thus, instead of setting the environment variable ARGOCD_CONTROLLER_REPLICAS representing the number of replicas, the number of replicas can be read directly from the number of healthy replicas of the application controller deployment.
```

Comment:
The application controller service account does not necessarily has access to all resources in the cluster. 
Some users install controller with only [argocd-application-controller-role](https://github.com/argoproj/argo-cd/blob/master/manifests/base/application-controller/argocd-application-controller-role.yaml) and use it to manage remote clusters only.

So I suggest changing [argocd-application-controller-role](https://github.com/argoproj/argo-cd/blob/master/manifests/base/application-controller/argocd-application-controller-role.yaml) and allowing controller inspect own deployment and find out a number of replicas.

Reply:
That's a good point. I will add this point to the document.